Expose and fix fallback versions

scaytrase opened the pull request:
exposes #116

Ocramius commented on test/PackageVersionsTest/E2EInstallerTest.php:
```diff
@@ -187,6 +187,51 @@ public function testRemovingPluginWithNoDevDoesNotAttemptToGenerateVersions() :
+                    ],
+                    [
+                        'type' => 'vcs',
+                        'url' => 'https://github.com/scaytrase/package-versions-main-stub.git',
```
Is there a way to include the package definitions manually? This test shouldn't rely on network access

scaytrase replied:
Sure, this is just a dirty implementation of the problem. I think I can to replace it with `path` or `artifact` generated in test
https://getcomposer.org/doc/05-repositories.md#path

scaytrase replied:
Please, check. I'm generating local zip artifacts now, refactored some code for it into reusable function

scaytrase replied:
I can do further refactorings, if you need them (i.e. refactor out logic on generating and validating  package version asserts)

Ocramius commented on test/PackageVersionsTest/E2EInstallerTest.php:
```diff
@@ -262,19 +316,25 @@ static function (SplFileInfo $file) use ($zip) {
+    /**
+     * @param mixed[] $composerJsonContent
+     */
+    private function createGenericArtifact(string $name, string $version, array $composerJsonContent): void
```
We could use `'path'` to force composer to perform a local (symlink) installation or such, instead of having to provide actual zip files.

Alternatively, the packages could be of `'type' => 'metapackage'`, which doesn't lead to any file installation (https://github.com/Roave/SecurityAdvisories/blob/666ba252853924887ac57dc9f66e6b6af78d5a76/composer.json#L3)

scaytrase replied:
I think, I can try `path`. Though I still don't get the idea what for

Ocramius replied:
I'm trying to remove even further components from the patch, that's all. Maintaining these tests, especially if many things are being moved around, is going to be a challenge in the years to come

scaytrase replied:
Sorry for long delay, got back to this issue. I stumbled upon a question, should I rewrite all zips (even original, which presents before my changes) or only new packages I test (so there will be both zip for package versions artifact and path for other packages)?

Ocramius replied:
No worries, I'm also not actively watching this issue ATM: my suggestion is to get to a point where no sources are needed at all, and only the `composer.lock` and `installed.json` are produced along with the generated versions.

Ocramius commented on test/PackageVersionsTest/E2EInstallerTest.php:
```diff
@@ -187,6 +187,60 @@ public function testRemovingPluginWithNoDevDoesNotAttemptToGenerateVersions() :
+                    ['packagist' => false],
+                    [
+                        'type' => 'artifact',
+                        'url' => $this->tempArtifact,
```
If I understand this correctly, the `tempArtifact` contains a full repository manifest here: maybe we should inline the JSON definition to show all existing versions of all packages in here

scaytrase replied:
IIUC `tempArtifact` is a foder ([path to folder actually](https://getcomposer.org/doc/05-repositories.md#artifact)) with zip artifacts. composer then chooses the most corrects artifact from present zips

Ocramius commented on test/PackageVersionsTest/E2EInstallerTest.php:
```diff
@@ -187,6 +187,60 @@ public function testRemovingPluginWithNoDevDoesNotAttemptToGenerateVersions() :
+                'name'         => 'package-versions/e2e-transitive',
+                'require'      => [
+                    'ocramius/package-versions' => '1.0.0',
+                    'infection/infection' => '^0.15.0',
```
If I understand this correctly, this configuration will lead to `4.0.0` for the symfony version being dumped?

scaytrase replied:
No, you can check previous build where debug output was echoed - it installs 5.0 version

https://travis-ci.org/scaytrase/PackageVersions/jobs/639786230#L344-L347

But package versions detects `v4.4.3@f5697ab4cb14a5deed7473819e63141bf5352c36`

https://travis-ci.org/scaytrase/PackageVersions/jobs/642543459#L306

Ocramius replied:
Oh, 4.4.3 is not even in the set of packages you have here @_@

scaytrase replied:
Yep, that the main evidence of the problem with fallback lock - it exist there

Ocramius commented on test/PackageVersionsTest/E2EInstallerTest.php:
```diff
@@ -289,6 +349,29 @@ private function writeComposerJsonFile(array $config, string $directory) : void
+
+        self::assertSame(0, $exitCode);
+        self::assertCount(1, $output);
+        self::assertRegExp('/^v?5\\..*\\@[a-f0-9]*$/', $output[0]);
```
Probably `self::assertStringStartsWith('v5.0.0', ...)`, since the rest of the regex is not relevant to this scenario

scaytrase replied:
Sounds valid, will fix that too

scaytrase commented:
> What are the generated composer.lock and installed.json before use-package-versions.php is run? Do they change afterwards?

Didn't get the question. You suggesting me to do some additional asserts before calling `use-package-versions.pgp` ?

Ocramius commented:
> Didn't get the question. You suggesting me to do some additional asserts before calling `use-package-versions.php` ?

No, I'm asking if you (with debugger or other tool) can inspect the state of `installed.php` before/after the plugin has run. The generated `Versions` file is built from `composer.lock` and `installed.json`, and those are either completely wrong, or wrong at a certain point in time during plugin run.

scaytrase commented:
Ok, will try to do it

scaytrase commented:
I've tried to dump files from the test itself using this patch

```patch
Index: test/PackageVersionsTest/E2EInstallerTest.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- test/PackageVersionsTest/E2EInstallerTest.php	(revision 2a189ff049d91162e8706d0ec824b5324732edce)
+++ test/PackageVersionsTest/E2EInstallerTest.php	(date 1581358608256)
@@ -237,8 +237,19 @@
             $this->tempLocalComposerHome . '/vendor/ocramius/package-versions/src/PackageVersions/Versions.php'
         );
 
+        copy($this->tempLocalComposerHome . '/composer.lock', __DIR__ . '/composer.lock.before');
+        copy($this->tempLocalComposerHome . '/vendor/composer/installed.json', __DIR__ . '/installed.json.before');
+        copy($this->tempLocalComposerHome . '/vendor/ocramius/package-versions/src/PackageVersions/Versions.php', __DIR__ . '/Versions.php.before');
+
         $this->writeSymfonyPackageVersionUsingFile($this->tempLocalComposerHome);
-        $this->assertSymfonyPackageVersionsIsUsable($this->tempLocalComposerHome);
+
+        try {
+            $this->assertSymfonyPackageVersionsIsUsable($this->tempLocalComposerHome);
+        } finally {
+            copy($this->tempLocalComposerHome . '/composer.lock', __DIR__ . '/composer.lock.after');
+            copy($this->tempLocalComposerHome . '/vendor/composer/installed.json', __DIR__ . '/installed.json.after');
+            copy($this->tempLocalComposerHome . '/vendor/ocramius/package-versions/src/PackageVersions/Versions.php', __DIR__ . '/Versions.php.after');
+        }
     }
 
     /**
```

Dumped files here:

[PackageVersionsTest.zip](https://github.com/Ocramius/PackageVersions/files/4182500/PackageVersionsTest.zip)

Shortly: they look correct and they are identical

actually, applying this patch fixes the test. maybe the matter is how FallbackVersions processes the fallback sources

```patch
Index: src/PackageVersions/FallbackVersions.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/PackageVersions/FallbackVersions.php	(revision 2a189ff049d91162e8706d0ec824b5324732edce)
+++ src/PackageVersions/FallbackVersions.php	(date 1581359614732)
@@ -59,14 +59,14 @@
     private static function getPackageData() : array
     {
         $checkedPaths = [
+            // This package's composer.lock
+            __DIR__ . '/../../composer.lock',
             // The top-level project's ./vendor/composer/installed.json
             getcwd() . '/vendor/composer/installed.json',
             __DIR__ . '/../../../../composer/installed.json',
             // The top-level project's ./composer.lock
             getcwd() . '/composer.lock',
             __DIR__ . '/../../../../../composer.lock',
-            // This package's composer.lock
-            __DIR__ . '/../../composer.lock',
         ];
 
         $packageData = [];
```

Ocramius commented:
I've lost track of this patch, sorry: anything else that needs to be done? CI still red, so I don't remember if we ever got to a resolution path.

scaytrase commented:
Patch fixing a test is mentioned above. I can apply it to this PR to push travis to go

Ocramius commented:
Gladly please! Feel free to remove any WIP markers too, and I'll try to make it go through after checking out locally once you pushed :+1:

scaytrase commented:
Hm, now only builds with lower or locked deps are failing, I'll check the composer.json

scaytrase commented:
@Ocramius I've fixed the tests but after I added reset clause for Versions.php to travis - psalm started complaining, take a look on it, please, looks not relevant to this PR. I can play with `travis.yml` a bit more to replay the previus conditins where psalm was passing (with pre-filled versions from composer install) but looks like the problem should be fixed elsewhere.

Plasm is complaining about impure calls if Versions.php is on it's original state (empty)

scaytrase commented:
Also for some reason travis is not reporting back to this PR. here is the build 

https://travis-ci.org/github/Ocramius/PackageVersions/builds/671962566

Ocramius commented:
Most likely superseded by #139

Ocramius commented:
I've done work to get rid of the entire dependency map in #146.

This is excellent work and investigation, @scaytrase, but I think the clear path forward at this point is to get rid with any lockfile/install map and to move everyone forward to `composer/composer:^2.0.0`.

I'm closing this, and will mark it as solved for `1.10.0`, which no longer presents the defect (the defect code is effectively no longer in use :D )

Thanks for working on this!
